pyogrio/_err.pyx

```diff
@@ -248,3 +339,121 @@ cdef void error_handler(CPLErr err_class, int err_no, const char* err_msg) nogil
+                    error.__cause__ = cause
+
+            if stack:
+                last = stack.pop()
+                if last is not None:
+                    raise last
+
+        return err
+
+    cdef void *exc_wrap_pointer(self, void *ptr) except NULL:
+        """Wrap a GDAL/OGR function that returns a pointer.
+
+        Raises an exception if a non-fatal error has been added to the
+        exception stack. Only checks for errors if `ptr` is `NULL`.
+        """
+        if ptr == NULL:
+            stack = self.error_stack.get()
+            for error, cause in zip_longest(stack[::-1], stack[::-1][1:]):
+                if error is not None and cause is not None:
+                    error.__cause__ = cause
+
+            if stack:
+                last = stack.pop()
+                if last is not None:
+                    raise last
+
+        return ptr
+
+
+cdef void stacking_error_handler(
+    CPLErr err_class,
+    int err_no,
+    const char* err_msg
+) noexcept with gil:
+    """Custom CPL error handler that adds non-fatal errors to a stack.
+
+    All non-fatal errors (CE_Failure) are not printed to stderr (behaviour
+    of the default GDAL error handler), but they are converted to python
+    exceptions and added to a stack, so they can be dealt with afterwards.
+
+    Warnings are converted to Python warnings.
+    """
+    global _ERROR_STACK
+
+    if err_class == CE_Fatal:
+        # If the error class is CE_Fatal, we want to have a message issued
+        # because the CPL support code does an abort() before any exception
+        # can be generated
+        CPLDefaultErrorHandler(err_class, err_no, err_msg)
+
+        return
+
+    elif err_class == CE_Failure:
+        # For Failures, add them to the error exception stack
+        stack = _ERROR_STACK.get()
+        stack.append(
+            exception_map.get(err_no, CPLE_BaseError)(
+                err_class, err_no, clean_error_message(err_msg)
+            ),
+        )
+        _ERROR_STACK.set(stack)
+
+        return
+
+    elif err_class == CE_Warning:
+        warnings.warn(clean_error_message(err_msg), RuntimeWarning)
+
+        return
+
+    # Fall back to the default handler for non-failure messages since
+    # they won't be translated into exceptions.
+    CPLDefaultErrorHandler(err_class, err_no, err_msg)
+
+
+@contextlib.contextmanager
+def capture_errors():
+    """A context manager that captures all GDAL non-fatal errors occuring.
+
+    It adds all errors to a single stack, so it assumes that no more than one
+    GDAL function is called.
+
+    Yields an ErrorHandler object that can be used to handle the errors
+    if any were captured.
+    """
+    CPLErrorReset()
+    global _ERROR_STACK
+    _ERROR_STACK.set([])
+
+    # stacking_error_handler records GDAL errors in the order they occur and
+    # converts them to exceptions.
+    CPLPushErrorHandlerEx(<CPLErrorHandler>stacking_error_handler, NULL)
```

Comment:
If we would (eventually?) handle all catching of GDAL errors using this context manager, then we don't need to each time register this error handler and pop it again?

Reply:
True. I'm not sure why both approaches (or even three taking the second context handler in account) are being used in rasterio/fiona...

In fiona at least it is only used when files are opened, all other calls use the getlasterror approach... possibly there is a downside to capturing all errors in some cases?

Reply:
> I'm not sure why both approaches are being used in rasterio/fiona...

That might also be partly just historical reasons, since this was only added more recently to fiona (and although it could be used everywhere, there was maybe also no pressing need to change the code that much?)